test/HistoryTest.cs

```diff
@@ -118,136 +119,289 @@ public void SensitiveHistoryOptionalBehavior()
+            {
                 options.HistorySavePath = oldHistoryFilePath;
                 options.HistorySaveStyle = oldHistorySaveStyle;
                 options.AddToHistoryHandler = PSConsoleReadLineOptions.DefaultAddToHistoryHandler;
-                File.Delete(newHistoryFilePath);
-            }
+                File.Delete(newHistoryFilePath);
+            }
+        }
+
+        [SkippableFact]
+        public void SensitiveHistoryOptionalBehaviorWithScriptBlock()
+        {
+            TestSetup(KeyMode.Cmd);
+
+            // No history
+            SetHistory();
+            Test("", Keys(_.UpArrow, _.DownArrow));
+
+            var options = PSConsoleReadLine.GetOptions();
+            var oldHistoryFilePath = options.HistorySavePath;
+            var oldHistorySaveStyle = options.HistorySaveStyle;
+
+            // AddToHistoryHandler should be set to the default handler.
+            Assert.Same(PSConsoleReadLineOptions.DefaultAddToHistoryHandler, options.AddToHistoryHandler);
+
+            var newHistoryFilePath = Path.GetTempFileName();
+            var newHistorySaveStyle = HistorySaveStyle.SaveIncrementally;
+            Func<string, object> newAddToHistoryHandler_ReturnBool = LanguagePrimitives.ConvertTo<Func<string, object>>(
+                ScriptBlock.Create(@"
+                    param([string]$line)
+                    $line.Contains('gal')"));
+            Func<string, object> newAddToHistoryHandler_ReturnEnum = LanguagePrimitives.ConvertTo<Func<string, object>>(
+                ScriptBlock.Create(@"
+                    param([string]$line)
+                    if ($line.Contains('gal')) {
+                        [Microsoft.PowerShell.AddToHistoryOption]::MemoryOnly
+                    } elseif ($line.Contains('gmo')) {
+                        'SkipAdding'
+                    } else {
+                        [Microsoft.PowerShell.AddToHistoryOption]::MemoryAndFile
+                    }"));
+            Func<string, object> newAddToHistoryHandler_ReturnOther = LanguagePrimitives.ConvertTo<Func<string, object>>(
+                ScriptBlock.Create(@"
+                    param([string]$line)
+                    'string value'"));
+
+            string[] commandInputs = new[] {
+                "gmo p*",
+                "gcm c*",
+                "gal dir",
+                "ConvertTo-SecureString -AsPlainText -String abc -Force"
+            };
+
+            string[] expectedQueuedItems = new[] {
+                "gcm c*",
+                "gal dir",
+                "ConvertTo-SecureString -AsPlainText -String abc -Force"
+            };
+
+            string[] expectedSavedItems = new[] {
+                "gcm c*",
+                "ConvertTo-SecureString -AsPlainText -String abc -Force"
+            };
+
+            try
+            {
+                options.HistorySavePath = newHistoryFilePath;
+                options.HistorySaveStyle = newHistorySaveStyle;
+
+                /*
```

Comment:
This project uses `//` for multi-line comments exclusively (modulo the copyright notices.)

Reply:
Will change to use `//`.